File: src/2025-02-27-deepclaude37-rust/javascript/exercises/practice/complex-numbers/complex-numbers.js
//
// This is only a SKELETON file for the 'Complex Numbers' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class ComplexNumber {
  constructor(a, b) {
    this.a = a;
    this.b = b;
    
    // Define conj as a non-enumerable property to prevent Jest from infinitely traversing it
    Object.defineProperty(this, 'conj', {
      get: function() {
        return new ComplexNumber(this.a, -this.b);
      },
      enumerable: false
    });
  }

  get real() {
    return this.a;
  }

  get imag() {
    return this.b;
  }

  add(other) {
    return new ComplexNumber(this.a + other.a, this.b + other.b);
  }

  sub(other) {
    return new ComplexNumber(this.a - other.a, this.b - other.b);
  }

  div(other) {
    const denom = other.a ** 2 + other.b ** 2;
    return new ComplexNumber(
      (this.a * other.a + this.b * other.b) / denom,
      (this.b * other.a - this.a * other.b) / denom
    );
  }

  mul(other) {
    return new ComplexNumber(
      this.a * other.a - this.b * other.b,
      this.b * other.a + this.a * other.b
    );
  }

  get abs() {
    return Math.sqrt(this.a ** 2 + this.b ** 2);
  }


  get exp() {
    const ea = Math.exp(this.a);
    return new ComplexNumber(
      ea * Math.cos(this.b),
      ea * Math.sin(this.b)
    );
  }
}

File: src/2025-02-27-deepclaude37-rust/cpp/exercises/practice/complex-numbers/complex_numbers.cpp
#include "complex_numbers.h"
#include <cmath>

namespace complex_numbers {

Complex::Complex(double real, double imag) : real_(real), imag_(imag) {}

Complex Complex::conj() const {
    return Complex(real_, -imag_);
}

double Complex::abs() const {
    return std::sqrt(real_ * real_ + imag_ * imag_);
}

Complex operator+(const Complex& lhs, const Complex& rhs) {
    return Complex(lhs.real() + rhs.real(), lhs.imag() + rhs.imag());
}

Complex operator-(const Complex& lhs, const Complex& rhs) {
    return Complex(lhs.real() - rhs.real(), lhs.imag() - rhs.imag());
}

Complex operator*(const Complex& lhs, const Complex& rhs) {
    double a = lhs.real();
    double b = lhs.imag();
    double c = rhs.real();
    double d = rhs.imag();
    return Complex(a * c - b * d, b * c + a * d);
}

Complex operator/(const Complex& lhs, const Complex& rhs) {
    double a = lhs.real();
    double b = lhs.imag();
    double c = rhs.real();
    double d = rhs.imag();
    double denominator = c * c + d * d;
    return Complex((a * c + b * d) / denominator, (b * c - a * d) / denominator);
}

Complex Complex::exp() const {
    double a = real_;
    double b = imag_;
    double ea = std::exp(a);
    return Complex(ea * std::cos(b), ea * std::sin(b));
}

}  // namespace complex_numbers
